### PR TITLE
Feat(TFD-174): Add Replicate processing-component

### PR DESCRIPTION
**Pull request by StefanLMNG**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [x]  Tests for the changes have been added (for bug fixes / features)

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

First implementation of the TCOMP Replicate component.

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:
Regarding which processing-component's PR is merged first, the ProcessingFamilyDefinition will be a bit changed to add the other definition components.

### CODE REVIEW

File: components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/replicate/replicateDefinition.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    public Class getPropertyClass() {
+        return ReplicateProperties.class;
+    }
+
+    public Property[] getReturnProperties() {
+        return new Property[] {};
+    }
+
+    @Override
+    public String getPngImagePath(ComponentImageType imageType) {
```
Comment (pbailly) [inline components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/replicate/replicateDefinition.java]: This method is the same as the one in its parent, you can remove it.

File: components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/replicate/replicateDefinition.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            return "Replicate_icon32.png"; //$NON-NLS-1$
+        default:
+            return "Replicate_icon32.png"; //$NON-NLS-1$
+        }
+    }
+
+    @Override
+    public RuntimeInfo getRuntimeInfo(ExecutionEngine engine, ComponentProperties properties,
+            ConnectorTopology connectorTopology) {
+        try {
```
Comment (pbailly) [inline components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/replicate/replicateDefinition.java]: Please add call to
`        assertEngineCompatibility(engine);`
and
`        assertConnectorTopologyCompatibility(connectorTopology);`

File: components/components-processing/processing-runtime/src/test/resources/org/talend/components/processing/runtime/ReplicateRuntimeTest.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        );
+
+        PCollection<IndexedRecord> input = (PCollection<IndexedRecord>) p.apply(Create.of(data).withCoder(LazyAvroCoder.of()));
+
+        ReplicateProperties replicateProperties = new ReplicateProperties("test");
+        replicateRuntime.initialize(null, replicateProperties);
+        BeamJobContext ctx = Mockito.mock(BeamJobContext.class);
+        when(ctx.getLinkNameByPortName(anyString())).thenReturn("test");
+        when(ctx.getPCollectionByLinkName(anyString())).thenReturn(input);
+        replicateRuntime.build(ctx);
```
Comment (pbailly) [inline components/components-processing/processing-runtime/src/test/resources/org/talend/components/processing/runtime/ReplicateRuntimeTest.java]: You should check your results.
For example, you could add a test that check that `ctx.getLinkNameByPortName()` and `ctx.getPCollectionByLinkName()` are called one time with:
`verify(ctx, once()).getLinkNameByPortName(anyString());`

File: components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/replicate/ReplicateProperties.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        }
+    };
+
+    // output schema
+    public transient PropertyPathConnector FLOW_CONNECTOR = new PropertyPathConnector(Connector.MAIN_NAME, "schemaFlow");
+
+    public transient PropertyPathConnector SECOND_FLOW_CONNECTOR = new PropertyPathConnector(Connector.MAIN_NAME,
+            "schemaSecondFlow");
+
+    public SchemaProperties schemaFlow = new SchemaProperties("schemaFlow");
```
Comment (pbailly) [inline components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/replicate/ReplicateProperties.java]: These schema are never used on the runtime, I'm not sure you have to define them here.

Comment (StefanLMNG) [inline components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/replicate/ReplicateProperties.java]: I will need to check with you on this point.

File: components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/replicate/replicateDefinition.java (hunk 1 above)
Comment (StefanLMNG) [inline components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/replicate/replicateDefinition.java]: Done !

File: components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/replicate/replicateDefinition.java (hunk 2 above)
Comment (StefanLMNG) [inline components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/replicate/replicateDefinition.java]: Done !

File: components/components-processing/processing-runtime/src/test/resources/org/talend/components/processing/runtime/ReplicateRuntimeTest.java (hunk 3 above)
Comment (StefanLMNG) [inline components/components-processing/processing-runtime/src/test/resources/org/talend/components/processing/runtime/ReplicateRuntimeTest.java]: Done !

File: components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/replicate/replicateDefinition.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import org.talend.components.api.component.runtime.DependenciesReader;
+import org.talend.components.api.component.runtime.ExecutionEngine;
+import org.talend.components.api.component.runtime.JarRuntimeInfo;
+import org.talend.components.api.exception.ComponentException;
+import org.talend.components.api.properties.ComponentProperties;
+import org.talend.components.processing.definition.ProcessingFamilyDefinition;
+import org.talend.daikon.properties.property.Property;
+import org.talend.daikon.runtime.RuntimeInfo;
+
+public class ReplicateDefinition extends AbstractComponentDefinition {
```
Comment (RyanSkraba) [inline components/components-processing/processing-definition/src/main/java/org/talend/components/processing/definition/replicate/replicateDefinition.java]: The file is `replicateDefinition.java`, which is the wrong case.

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/520/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/520/)

**Comment by build-talend-common** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/538/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/538/)

**Pull request merged**